File: handyvertrag-app/scripts/parse-feeds.py
#!/usr/bin/env python3
"""
BELLA Feed-Parser — liest AWIN- + AdCell-Feeds, filtert HUNDE-FUTTER (kein Zubehör,
keine Katze), normalisiert auf das dog_foods-Schema und schreibt dog_foods.json.

Run:  python scripts/parse-feeds.py
Danach:  node scripts/load-dog-foods.mjs   (braucht DATABASE_URL)
"""
import csv, gzip, re, io, json, sys, os

csv.field_size_limit(10**7)

# ── Feed-Quellen (anpassen wenn neue Feeds kommen) ───────────────────────────
DL = os.path.expanduser("~/Downloads")
AWIN_FEEDS = [os.path.join(DL, "11703-23513-de_DE-Default.csv.gz")]   # schecker.de
ADCELL_FEEDS = [
    os.path.join(DL, "419197-66376.csv"),
    os.path.join(DL, "521034-66376.csv"),
    os.path.join(DL, "496158-66376.csv"),
]
PUBLISHER = "615299"  # AWIN publisher id (für Tracking-Konsistenz)

# ── Filter-Heuristiken ───────────────────────────────────────────────────────
FOOD_RE = re.compile(r"futter|nahrung|men[üu]\b|trockenfutter|nassfutter|kausnack|kauknochen|kaustange|kauartikel|leckerli|trainingssnack|barf|frostfutter|frischfleisch|dose|dosen|nass\b|trocken\b|kroketten|flocken|alleinfutter|erg[äa]nzungsfutter", re.I)
ACCESSORY_RE = re.compile(r"beutel|tasche|pocket|napf|leine|halsband|geschirr|spielzeug|\bball\b|decke|k[öo]rbchen|kissen|b[üu]rste|\bkamm\b|shampoo|mantel|jacke|schuhe|pfote(?:n)?schutz|transportbox|\bbox\b|eimer|zaun|gitter|clicker|pfeife|markierungs|kotbeutel|handschuh|h[üu]rde|tunnel|g[üu]rtel|rucksack|tragetasche|buch|dvd", re.I)
DOG_RE = re.compile(r"hund|dog|welpe|barf|doggy", re.I)
OTHER_PET_RE = re.compile(r"katz|\bcat\b|nager|kaninchen.?stall|vogel|\bfisch.?tank|aquarium|pferd|reit", re.I)

def is_dog_food(title, cat, merchant_is_dog=False):
    blob = f"{title} {cat}".lower()
    if ACCESSORY_RE.search(blob):
        return False
    if OTHER_PET_RE.search(blob) and not DOG_RE.search(blob):
        return False
    if not (merchant_is_dog or DOG_RE.search(blob)):
        return False
    return bool(FOOD_RE.search(blob))

def infer_type(t):
    t = t.lower()
    if re.search(r"kausnack|kauknochen|kaustange|kauartikel|leckerli|trainingssnack|\bsnack|knochen|sticks?\b", t): return "snack"
    if re.search(r"barf|frostfutter|frischfleisch|\broh\b", t): return "barf"
    if re.search(r"nassfutter|nass-|\bdose|dosen|frischebeutel|men[üu]\b|feucht|pastete", t): return "nass"
    if re.search(r"kaltgepresst", t): return "kaltgepresst"
    if re.search(r"trockenfutter|trocken|kroketten|flocken", t): return "trocken"
    return "trocken"

PROTEINS = [("h[äa]hnchen","Huhn"),("huhn","Huhn"),("rind","Rind"),("lachs","Lachs"),("lamm","Lamm"),
            ("ente","Ente"),("pute","Pute"),("truthahn","Pute"),("wild","Wild"),("kaninchen","Kaninchen"),
            ("pferd","Pferd"),("ziege","Ziege"),("insekt","Insekt"),("fisch","Fisch"),("strauss","Strauß")]
def infer_protein(t):
    t = t.lower()
    for k, lab in PROTEINS:
        if re.search(k, t): return lab
    return None

def suitable_for(t):
    t = t.lower(); s = []
    if re.search(r"welpe|junior|puppy", t): s.append("welpen")
    if re.search(r"senior|\balt\b|ageing|aging|7\+|8\+", t): s.append("senior")
    if re.search(r"adult|erwachsen", t): s.append("adult")
    if re.search(r"allergie|sensitiv|hypoallergen|getreidefrei|grain.?free|monoprotein|magen", t): s.append("allergie")
    return s or ["adult"]

def num(s):
    if not s: return None
    s = str(s).strip()
    # "12,99" / "1.234,56" / "12.99"
    if "," in s and "." in s: s = s.replace(".", "").replace(",", ".")
    else: s = s.replace(",", ".")
    m = re.search(r"-?\d+(?:\.\d+)?", s)
    return float(m.group(0)) if m else None

def weight_kg(t):
    t = t.lower()
    m = re.search(r"(\d+(?:[.,]\d+)?)\s*kg\b", t)
    if m: return float(m.group(1).replace(",", "."))
    m = re.search(r"(\d+)\s*g\b", t)
    if m: return float(m.group(1)) / 1000
    return None

def slugify(s):
    return re.sub(r"[^a-z0-9]+", "-", (s or "").lower()).strip("-")[:90]

def smart_open(path):
    """Decode cp1252 (Latin-1) feeds correctly; fall back to utf-8."""
    raw = gzip.open(path, "rb").read() if path.endswith(".gz") else open(path, "rb").read()
    try:
        return io.StringIO(raw.decode("utf-8"))
    except UnicodeDecodeError:
        return io.StringIO(raw.decode("cp1252", errors="replace"))

# ── Parsing ──────────────────────────────────────────────────────────────────
out = {}  # slug -> record (dedupe)

def add(rec):
    if not rec["affiliateUrl"] or not rec["name"]:
        return
    out.setdefault(rec["slug"], rec)

def parse_awin(path):
    n = 0
    merchant_is_dog = "schecker" in path.lower()
    for row in csv.DictReader(smart_open(path)):
        title = row.get("product_name") or ""
        cat = row.get("merchant_category") or ""
        if not is_dog_food(title, cat, merchant_is_dog):
            continue
        if (cat.split(">")[0] not in ("Hundefutter", "Hundesnacks")) and merchant_is_dog:
            continue
        price = num(row.get("search_price"))
        wk = weight_kg(title)
        ppk = round(price / wk, 2) if (price and wk and wk > 0) else None
        brand = (title.split()[0] if title else "").strip("®™,")
        add({
            "slug": slugify(f"{brand}-{title}-{row.get('aw_product_id','')}"),
            "brand": brand or row.get("merchant_name") or "",
            "name": title.strip(),
            "type": infer_type(f"{title} {cat}"),
            "protein": infer_protein(title),
            "isGrainFree": bool(re.search(r"getreidefrei|grain.?free", title, re.I)),
            "isHypoallergenic": bool(re.search(r"hypoallergen|sensitiv|monoprotein|allergie", title, re.I)),
            "pricePerKg": ppk,
            "price": price,
            "suitableFor": suitable_for(f"{title} {cat}"),
            "imageUrl": row.get("merchant_image_url") or row.get("aw_image_url") or None,
            "affiliateNetwork": "awin",
            "affiliateUrl": row.get("aw_deep_link") or row.get("merchant_deep_link") or "",
            "source": row.get("merchant_name") or "awin",
        })
        n += 1
    return n

def parse_adcell(path):
    n = 0
    for row in csv.DictReader(smart_open(path), delimiter=";"):
        title = row.get("Produkt-Titel") or ""
        cat = row.get("Produktkategorie") or ""
        if not is_dog_food(title, cat):
            continue
        price = num(row.get("Preis (Brutto)"))
        gp_unit = (row.get("Grundpreiseinheit") or "").lower()
        ppk = num(row.get("Grundpreis")) if ("kg" in gp_unit) else None
        if ppk is None:
            wk = weight_kg(f"{title} {row.get('Inhalt','')}")
            ppk = round(price / wk, 2) if (price and wk and wk > 0) else None
        brand = (row.get("Hersteller") or (title.split()[0] if title else "")).strip("®™,")
        add({
            "slug": slugify(f"{brand}-{title}-{row.get('europäische Artikelnummer EAN','') or row.get('Anbieter Artikelnummer AAN','')}"),
            "brand": brand,
            "name": title.strip(),
            "type": infer_type(f"{title} {cat}"),
            "protein": infer_protein(title),
            "isGrainFree": bool(re.search(r"getreidefrei|grain.?free", title, re.I)),
            "isHypoallergenic": bool(re.search(r"hypoallergen|sensitiv|monoprotein|allergie", title, re.I)),
            "pricePerKg": ppk,
            "price": price,
            "suitableFor": suitable_for(f"{title} {cat}"),
            "imageUrl": row.get("Produktbild-URL") or row.get("Vorschaubild-URL") or None,
            "affiliateNetwork": "adcell",
            "affiliateUrl": row.get("Deeplink") or "",
            "source": brand or "adcell",
        })
        n += 1
    return n

print("BELLA Feed-Parser")
for p in AWIN_FEEDS:
    if os.path.exists(p): print(f"  AWIN {os.path.basename(p)}: {parse_awin(p)} Futter")
for p in ADCELL_FEEDS:
    if os.path.exists(p): print(f"  AdCell {os.path.basename(p)}: {parse_adcell(p)} Futter")

records = list(out.values())
json.dump(records, open(os.path.join(os.path.dirname(__file__), "dog_foods.json"), "w", encoding="utf-8"), ensure_ascii=False)
print(f"\n  -> {len(records)} eindeutige Hunde-Futter-Produkte in scripts/dog_foods.json")
print("\n  Typ-Verteilung:", {t: sum(1 for r in records if r['type']==t) for t in set(r['type'] for r in records)})
print("  mit €/kg:", sum(1 for r in records if r['pricePerKg']), "| mit Protein:", sum(1 for r in records if r['protein']))
print("\n  Beispiele:")
for r in records[:6]:
    print(f"   - {r['name'][:50]} | {r['brand']} | {r['type']} | {r['protein']} | {r['pricePerKg']}€/kg")

File: bella-app/scripts/lib/parse-feeds.mjs
/**
 * BELLA Feed-Parser (Node-Port von scripts/parse-feeds.py) — liest AWIN- + AdCell-Feeds,
 * filtert HUNDE-FUTTER (kein Zubehör, keine Katze), normalisiert auf das dog_foods-Schema.
 *
 * Feeds kommen aus AWIN_FEED_URLS / ADCELL_FEED_URLS (kommagetrennt) — Cron/Netlify-Function.
 * Ohne diese Env-Vars: lokaler Fallback auf Dateien in ~/Downloads (manuelle Runs).
 */
import { existsSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import path from "node:path";
import zlib from "node:zlib";
import { parse } from "csv-parse/sync";

// ── Filter-Heuristiken ───────────────────────────────────────────────────────
const FOOD_RE = /futter|nahrung|men[üu]\b|trockenfutter|nassfutter|kausnack|kauknochen|kaustange|kauartikel|leckerli|leckerchen|trainingssnack|barf|frostfutter|frischfleisch|dose|dosen|nass\b|trocken\b|kroketten|flocken|alleinfutter|erg[äa]nzungsfutter/i;
// Futter-Unterkategorien bei Mischshops (z.B. "Hund > BARF > Fertigbarf") — Pflege/Ergänzung ausgeschlossen
const FOOD_SUBCAT_RE = /^(BARF|Nassfutter|Trockenfutter|Snacks|Schonend gegart)/i;
const ACCESSORY_RE = /beutel|tasche|pocket|napf|leine|halsband|geschirr|spielzeug|\bball\b|decke|k[öo]rbchen|kissen|b[üu]rste|\bkamm\b|shampoo|mantel|jacke|schuhe|pfote(?:n)?schutz|transportbox|\bbox\b|eimer|zaun|gitter|clicker|pfeife|markierungs|kotbeutel|handschuh|h[üu]rde|tunnel|g[üu]rtel|rucksack|tragetasche|buch|dvd/i;
const DOG_RE = /hund|dog|welpe|barf|doggy/i;
const OTHER_PET_RE = /katz|\bcat\b|nager|kaninchen.?stall|vogel|\bfisch.?tank|aquarium|pferd|reit/i;

function isDogFood(title, cat, merchantIsDog = false) {
  const blob = `${title} ${cat}`.toLowerCase();
  if (ACCESSORY_RE.test(blob)) return false;
  if (OTHER_PET_RE.test(blob) && !DOG_RE.test(blob)) return false;
  if (!(merchantIsDog || DOG_RE.test(blob))) return false;
  return FOOD_RE.test(blob);
}

function inferType(t) {
  t = t.toLowerCase();
  if (/kausnack|kauknochen|kaustange|kauartikel|leckerli|trainingssnack|\bsnack|knochen|kn[öo]chelchen|sticks?\b|h[üu]lse|ohren|ziemer|smoothie|hundeeis/.test(t)) return "snack";
  if (/barf|frostfutter|frischfleisch|\broh\b/.test(t)) return "barf";
  if (/nassfutter|nass-|\bdose|dosen|frischebeutel|men[üu]\b|feucht|pastete|schonend gegart|frischemen/.test(t)) return "nass";
  if (/kaltgepresst/.test(t)) return "kaltgepresst";
  if (/trockenfutter|trocken|kroketten|flocken/.test(t)) return "trocken";
  return "trocken";
}

const PROTEINS = [
  ["h[äa]hnchen", "Huhn"], ["huhn", "Huhn"], ["h[üu]hn", "Huhn"], ["rind", "Rind"], ["lachs", "Lachs"], ["lamm", "Lamm"],
  ["ente", "Ente"], ["pute", "Pute"], ["truthahn", "Pute"], ["wild", "Wild"], ["kaninchen", "Kaninchen"],
  ["pferd", "Pferd"], ["ziege", "Ziege"], ["insekt", "Insekt"], ["fisch", "Fisch"], ["strauss", "Strauß"],
];
function inferProtein(t) {
  t = (t || "").toLowerCase();
  for (const [k, lab] of PROTEINS) {
    if (new RegExp(k).test(t)) return lab;
  }
  return null;
}

function suitableFor(t) {
  t = (t || "").toLowerCase();
  const s = [];
  if (/welpe|junior|puppy/.test(t)) s.push("welpen");
  if (/senior|\balt\b|ageing|aging|7\+|8\+/.test(t)) s.push("senior");
  if (/adult|erwachsen/.test(t)) s.push("adult");
  if (/allergie|sensitiv|hypoallergen|getreidefrei|grain.?free|monoprotein|magen/.test(t)) s.push("allergie");
  return s.length ? s : ["adult"];
}

function num(s) {
  if (!s) return null;
  s = String(s).trim();
  if (s.includes(",") && s.includes(".")) s = s.replaceAll(".", "").replaceAll(",", ".");
  else s = s.replaceAll(",", ".");
  const m = s.match(/-?\d+(?:\.\d+)?/);
  return m ? parseFloat(m[0]) : null;
}

function weightKg(t) {
  t = (t || "").toLowerCase();
  let m = t.match(/(\d+(?:[.,]\d+)?)\s*kg\b/);
  if (m) return parseFloat(m[1].replace(",", "."));
  m = t.match(/(\d+)\s*g\b/);
  if (m) return parseFloat(m[1]) / 1000;
  return null;
}

function slugify(s) {
  return (s || "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 90);
}

function cleanTitle(title) {
  if (title.includes("|")) {
    let parts = title.split("|").map((p) => p.trim());
    parts = parts.filter((p) => p && p.toLowerCase() !== "haustierkost");
    if (parts.length) return parts[0];
  }
  return title.trim();
}

function stripBrandChars(s) {
  return (s || "").replace(/^[®™,]+|[®™,]+$/g, "");
}

function smartOpen(buf) {
  if (buf.length >= 2 && buf[0] === 0x1f && buf[1] === 0x8b) buf = zlib.gunzipSync(buf);
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(buf);
  } catch {
    return new TextDecoder("windows-1252").decode(buf);
  }
}

function add(out, rec) {
  if (!rec.affiliateUrl || !rec.name) return;
  if (!(rec.slug in out)) out[rec.slug] = rec;
}

function csvRows(content, delimiter) {
  return parse(content, { columns: true, delimiter, bom: true, skip_empty_lines: true, relax_column_count: true });
}

const GRAIN_FREE_RE = /getreidefrei|grain.?free/i;
const HYPOALLERGENIC_RE = /hypoallergen|sensitiv|monoprotein|allergie/i;

// Reine Hundefutter-Marken im AdCell-Netzwerk — Titel enthält oft kein Wort wie "Hund"
// (z.B. "Geflügel, Reis & Gemüse"), daher würde isDogFood() sie sonst verwerfen.
// Domain-Erkennung statt Modus-Parameter, damit das für ENV-URLs UND den lokalen
// Fallback gleich funktioniert, ohne das ENV-Variablen-Format ändern zu müssen.
const DOG_MERCHANT_DOMAINS = ["milo-mia.de", "salingo.de", "paulis-petfood.de"];
function isDogMerchantDeeplink(deeplink) {
  if (!deeplink) return false;
  try {
    const url = new URL(deeplink);
    const target = url.hostname.endsWith("adcell.com") ? url.searchParams.get("param0") : deeplink;
    if (!target) return false;
    const host = new URL(target).hostname.replace(/^www\./, "");
    return DOG_MERCHANT_DOMAINS.some((d) => host === d || host.endsWith(`.${d}`));
  } catch {
    return false;
  }
}

/**
 * mode:
 * - "schecker": Hunde-Shop, Top-Kategorie muss Hundefutter/Hundesnacks sein + isDogFood
 * - "petfood_brand": reiner Tiernahrung-Hersteller (z.B. bosch) — alles unter Top-Kat "Hund" ist Futter
 * - "mixed_shop": Mischshop (z.B. Haustierkost) — nur Futter-Subkategorien (BARF/Nass/Trocken/Snacks)
 * - "default": generische Heuristik (isDogFood)
 */
function parseAwin(content, mode, out) {
  let n = 0;
  for (const row of csvRows(content, ",")) {
    const title = cleanTitle(row.product_name || "");
    const cat = row.merchant_category || "";
    const parts = cat.split(">").map((p) => p.trim());
    const top = parts[0] || "";
    const sub = parts[1] || "";

    if (mode === "schecker") {
      if (!isDogFood(title, cat, true)) continue;
      if (!["Hundefutter", "Hundesnacks"].includes(top)) continue;
    } else if (mode === "petfood_brand") {
      if (top !== "Hund") continue;
    } else if (mode === "mixed_shop") {
      if (!["Hund", "Hund & Katze"].includes(top)) continue;
      if (!FOOD_SUBCAT_RE.test(sub)) continue;
    } else {
      if (!isDogFood(title, cat)) continue;
    }

    const price = num(row.search_price);
    const wk = weightKg(title);
    const ppk = price && wk && wk > 0 ? Math.round((price / wk) * 100) / 100 : null;
    const brand = stripBrandChars(title ? title.split(/\s+/)[0] : "");
    add(out, {
      slug: slugify(`${brand}-${title}-${row.aw_product_id || ""}`),
      brand: brand || row.merchant_name || "",
      name: title.trim(),
      type: inferType(`${title} ${cat}`),
      protein: inferProtein(title),
      isGrainFree: GRAIN_FREE_RE.test(title),
      isHypoallergenic: HYPOALLERGENIC_RE.test(title),
      pricePerKg: ppk,
      price,
      suitableFor: suitableFor(`${title} ${cat}`),
      imageUrl: row.merchant_image_url || row.aw_image_url || null,
      affiliateNetwork: "awin",
      affiliateUrl: row.aw_deep_link || row.merchant_deep_link || "",
      source: row.merchant_name || "awin",
    });
    n++;
  }
  return n;
}

function parseAdcell(content, out) {
  let n = 0;
  for (const row of csvRows(content, ";")) {
    const title = row["Produkt-Titel"] || "";
    const cat = row["Produktkategorie"] || "";
    const dogMerchant = isDogMerchantDeeplink(row["Deeplink"]);
    if (!isDogFood(title, cat, dogMerchant)) continue;
    const price = num(row["Preis (Brutto)"]);
    const gpUnit = (row["Grundpreiseinheit"] || "").toLowerCase();
    let ppk = gpUnit.includes("kg") ? num(row["Grundpreis"]) : null;
    if (ppk == null) {
      const wk = weightKg(`${title} ${row["Inhalt"] || ""}`);
      ppk = price && wk && wk > 0 ? Math.round((price / wk) * 100) / 100 : null;
    }
    const brand = stripBrandChars(row["Hersteller"] || (title ? title.split(/\s+/)[0] : ""));
    add(out, {
      slug: slugify(`${brand}-${title}-${row["europäische Artikelnummer EAN"] || row["Anbieter Artikelnummer AAN"] || ""}`),
      brand,
      name: title.trim(),
      type: inferType(`${title} ${cat}`),
      protein: inferProtein(title),
      isGrainFree: GRAIN_FREE_RE.test(title),
      isHypoallergenic: HYPOALLERGENIC_RE.test(title),
      pricePerKg: ppk,
      price,
      suitableFor: suitableFor(`${title} ${cat}`),
      imageUrl: row["Produktbild-URL"] || row["Vorschaubild-URL"] || null,
      affiliateNetwork: "adcell",
      affiliateUrl: row["Deeplink"] || "",
      source: brand || "adcell",
    });
    n++;
  }
  return n;
}

/**
 * AWIN-Feed im Google-Shopping-Format (title/price/brand/image_link/aw_deep_link).
 * Wird für reine Snack-/Kauartikel-Shops genutzt (z.B. fidelis.dog).
 */
function parseAwinShopping(content, out) {
  let n = 0;
  for (const row of csvRows(content, ",")) {
    const title = row.title || "";
    const price = num(row.price);
    const wk = weightKg(title);
    const ppk = price && wk && wk > 0 ? Math.round((price / wk) * 100) / 100 : null;
    const brand = stripBrandChars(row.brand || (title ? title.split(/\s+/)[0] : ""));
    add(out, {
      slug: slugify(`${brand}-${title}-${row.gtin || row.mpn || ""}`),
      brand,
      name: title.trim(),
      type: inferType(title),
      protein: inferProtein(title),
      isGrainFree: GRAIN_FREE_RE.test(title),
      isHypoallergenic: HYPOALLERGENIC_RE.test(title),
      pricePerKg: ppk,
      price,
      suitableFor: suitableFor(title),
      imageUrl: row.image_link || null,
      affiliateNetwork: "awin",
      affiliateUrl: row.aw_deep_link || row.link || "",
      source: brand || "awin",
    });
    n++;
  }
  return n;
}

/**
 * AWIN liefert zwei unterschiedliche CSV-Layouts unter demselben "AWIN_FEED_URLS":
 * das Standard-Produktfeed (product_name/merchant_category/search_price) und das
 * Google-Shopping-Format (title/price/brand, z.B. Mera Tiernahrung). Header-Zeile
 * prüfen statt einen festen Modus zu erzwingen, sonst landen Shopping-Feeds als
 * 0-Treffer-Fehlschlag (Spalten existieren schlicht nicht).
 */
function detectAwinFormat(content) {
  const header = content.slice(0, content.indexOf("\n"));
  return header.includes("product_name") ? "default" : "shopping";
}

async function downloadFeed(url) {
  const res = await fetch(url, {
    headers: { "User-Agent": "BELLA-FeedBot/1.0" },
    signal: AbortSignal.timeout(120_000),
  });
  if (!res.ok) throw new Error(`Download fehlgeschlagen (${res.status}): ${url}`);
  return Buffer.from(await res.arrayBuffer());
}

export async function parseFeeds() {
  const out = {};
  console.log("BELLA Feed-Parser (Node)");

  // AWIN_FEED_URLS/ADCELL_FEED_URLS sind in Netlify als "sensitive" markiert — der
  // bestehende Wert lässt sich nicht mehr auslesen (auch nicht per CLI/Dashboard für
  // den Account-Owner), Anhängen würde also blind den unbekannten Altwert riskieren.
  // Neue Feeds kommen stattdessen über _EXTRA-Variablen, die mit dem Hauptwert
  // zusammengeführt werden — nicht-destruktiv, alter Wert bleibt unberührt.
  const splitUrls = (v) => (v || "").split(",").map((s) => s.trim()).filter(Boolean);
  const awinUrls = [...splitUrls(process.env.AWIN_FEED_URLS), ...splitUrls(process.env.AWIN_FEED_URLS_EXTRA)];
  const adcellUrls = [...splitUrls(process.env.ADCELL_FEED_URLS), ...splitUrls(process.env.ADCELL_FEED_URLS_EXTRA)];

  if (awinUrls.length || adcellUrls.length) {
    for (const [i, url] of awinUrls.entries()) {
      console.log(`  ⬇ AWIN ${i}: ${url.slice(0, 70)}…`);
      const content = smartOpen(await downloadFeed(url));
      const format = detectAwinFormat(content);
      const n = format === "shopping" ? parseAwinShopping(content, out) : parseAwin(content, "default", out);
      console.log(`  AWIN feed ${i} [${format}]: ${n} Futter`);
    }
    for (const [i, url] of adcellUrls.entries()) {
      console.log(`  ⬇ AdCell ${i}: ${url.slice(0, 70)}…`);
      const content = smartOpen(await downloadFeed(url));
      console.log(`  AdCell feed ${i}: ${parseAdcell(content, out)} Futter`);
    }
  } else {
    // Lokaler Fallback für manuelle Runs (Feed-Dateien in ~/Downloads)
    const dl = process.env.FEED_DIR || path.join(homedir(), "Downloads");
    const AWIN_FEEDS = [
      [path.join(dl, "11703-23513-de_DE-Default.csv.gz"), "schecker"],
      [path.join(dl, "56633-107909-de_DE-Default.csv.gz"), "default"],
      [path.join(dl, "84955-99728-de_DE-Produktdatenfeed.csv.gz"), "petfood_brand"],
      [path.join(dl, "datafeed_615299 (24).csv.gz"), "mixed_shop"],
    ];
    const ADCELL_FEEDS = [
      "419197-66376.csv", "521034-66376.csv", "496158-66376.csv", "630262-66376.csv",
      "540252-66376.csv", "372544-66376.csv", "434708-66376.csv",
    ].map((f) => path.join(dl, f));
    const AWIN_SHOPPING_FEEDS = [
      path.join(dl, "116601-retail-de_DE.csv.gz"),
      path.join(dl, "115623-retail-de_DE.csv.gz"),
    ];

    for (const [p, mode] of AWIN_FEEDS) {
      if (existsSync(p)) {
        const content = smartOpen(readFileSync(p));
        console.log(`  AWIN ${path.basename(p)} [${mode}]: ${parseAwin(content, mode, out)} Futter`);
      }
    }
    for (const p of ADCELL_FEEDS) {
      if (existsSync(p)) {
        const content = smartOpen(readFileSync(p));
        console.log(`  AdCell ${path.basename(p)}: ${parseAdcell(content, out)} Futter`);
      }
    }
    for (const p of AWIN_SHOPPING_FEEDS) {
      if (existsSync(p)) {
        const content = smartOpen(readFileSync(p));
        console.log(`  AWIN-Shopping ${path.basename(p)}: ${parseAwinShopping(content, out)} Futter`);
      }
    }
  }

  const records = Object.values(out);
  console.log(`\n  -> ${records.length} eindeutige Hunde-Futter-Produkte`);
  const typeDist = {};
  for (const r of records) typeDist[r.type] = (typeDist[r.type] || 0) + 1;
  console.log("  Typ-Verteilung:", typeDist);
  console.log("  mit €/kg:", records.filter((r) => r.pricePerKg).length, "| mit Protein:", records.filter((r) => r.protein).length);
  console.log("\n  Beispiele:");
  for (const r of records.slice(0, 6)) {
    console.log(`   - ${r.name.slice(0, 50)} | ${r.brand} | ${r.type} | ${r.protein} | ${r.pricePerKg}€/kg`);
  }
  return records;
}
